src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs

```diff
@@ -415,10 +418,22 @@ internal SynthesizedAttributeData SynthesizeDebuggerStepThroughAttribute()
             return TrySynthesizeAttribute(WellKnownMember.System_Diagnostics_DebuggerStepThroughAttribute__ctor);
         }
 
-        internal SynthesizedAttributeData SynthesizeReadOnlyAttribute()
+        internal void EnsureIsReadOnlyAttributeExists(Symbol readOnlySymbol)
         {
-            // PROTOTYPE(readonlyRefs) it is optional now as it will be generated in the next PR
-            return TrySynthesizeAttribute(WellKnownMember.System_Runtime_CompilerServices_ReadOnlyAttribute__ctor, isOptionalUse: true);
+            // This will report a DeclarationDiagnostics error if the member is not found.
+            // We only need to report this when compiling NetModules, as for other compilations, we will generate it.
+
+            var attributeSymbol = Binder.GetWellKnownTypeMember(
+                compilation: this,
+                member: WellKnownMember.System_Runtime_CompilerServices_IsReadOnlyAttribute__ctor,
+                diagnostics: DeclarationDiagnostics,
+                location: readOnlySymbol.Locations.FirstOrDefault() ?? Location.None,
+                isOptional: Options.OutputKind != OutputKind.NetModule);
+
+            if (attributeSymbol == null)
+            {
+                NeedsGeneratedIsReadOnlyAttribute = true;
+            }
```

Comment:
Can we do this check for `GetWellKnownTypeMember(IsReadOnlyAttribute_ctor)` at most once, at emit time, if there are any symbols that require it?